runtime/src/serde_snapshot.rs

```diff
@@ -321,19 +351,64 @@ where
     let mut accounts_db = AccountsDb::new_with_config(
         account_paths.to_vec(),
         cluster_type,
-        account_indexes,
+        account_secondary_indexes,
         caching_enabled,
         shrink_ratio,
     );
-    let AccountsDbFields(storage, version, slot, bank_hash_info) = accounts_db_fields;
+    let AccountsDbFields(
+        full_snapshot_storage,
+        full_snapshot_version,
+        full_snapshot_slot,
+        full_snapshot_bank_hash_info,
+    ) = snapshot_accounts_db_fields.full_snapshot_accounts_db_fields;
+
+    let (snapshot_version, snapshot_slot, snapshot_bank_hash_info) = snapshot_accounts_db_fields
+        .incremental_snapshot_accounts_db_fields
+        .as_ref()
+        .map(
+            |AccountsDbFields(
+                _,
+                incremental_snapshot_version,
+                incremental_snapshot_slot,
+                incremental_snapshot_bank_hash_info,
+            )| {
+                (
+                    *incremental_snapshot_version,
+                    *incremental_snapshot_slot,
+                    incremental_snapshot_bank_hash_info.clone(),
+                )
+            },
+        )
+        .unwrap_or((
+            full_snapshot_version,
+            full_snapshot_slot,
+            full_snapshot_bank_hash_info,
+        ));
+
+    // filter out incremental snapshot storages with slot <= full snapshot slot
+    let incremental_snapshot_storage = snapshot_accounts_db_fields
+        .incremental_snapshot_accounts_db_fields
+        .map_or_else(HashMap::new, |fields| {
+            fields
+                .0
+                .into_iter()
+                .filter(|(slot, _)| *slot > full_snapshot_slot)
+                .collect::<HashMap<_, _>>()
+        });
```

Comment:
Should we filter here or just return an error? Since we checked earlier via `check_are_snapshots_compatible` that the incremental snapshot was based on this full snapshot, there should be no slots `<= full snapshot slot` right?

Reply:
Yeah... this one confused me too. So when AccountsHashVerifier is taking hashes, it's doing it from snapshots. Those snapshots (not the whole snapshot archive) need to have all the storages to compute the hash correctly.

When I was writing tests in `core/tests/snapshots.rs`, which calls functions a little differently to create snapshot archives, I was running into issues where rebuild would fail due to AppendVec ID mismatch for slots less than the full snapshot slot in the incremental snapshot. The accounts_db_fields.0 also has info for all the slots here, so this filtering needs to happen when rebuilding with an incremental snapshot.

Both the accounts_db_fields.0 _and_ the actual storages need to be filtered out. The storages can get filtered out when creating the incremental snapshot archive, and then the accounts_db_fields.0 needs to be filtered out when rebuilding. Here's some more info when I was asking about it on Discord [1].

[1] https://discord.com/channels/428295358100013066/439194979856809985/854856681082978304

Reply:
Little confused here, let's have a brief chat to go over it.

Reply:
After talking with @carllin, I've created a new issue to track this: #18813. The gist is during serialization of accounts_db for an incremental snapshot, to filter _out_ the storage entries for all slots less-than-or-equal-to the full snapshot slot.

Reply:
Thanks!